docs/guide.md

```diff
@@ -0,0 +1,575 @@
+refers to the value of the string in this case.
+
+This example compiles to:
+
+```JSON
+{
+  "rules": {
+    "users": {
+      "$id": {
+        ".validate": "newData.hasChildren(['name', 'isAdmin'])",
+        "name": {
+          ".validate": "newData.isString() && newData.val().length > 0 && newData.val().length <= 32"
+        },
+        "isAdmin": {
+          ".validate": "newData.isBoolean()"
+        },
+        "$other": {
+          ".validate": "false"
+        }
+      }
+    },
+    "rooms": {
+      "$id": {
+        ".validate": "newData.hasChildren(['name', 'creator'])",
+        "name": {
+          ".validate": "newData.isString() && newData.val().length > 0 && newData.val().length <= 32"
+        },
+        "creator": {
+          ".validate": "newData.isString()"
+        },
+        "$other": {
+          ".validate": "false"
+        }
+      }
+    }
+  }
+}
+```
+
+## Functions
+
+Bolt also allows you to organize common expressions as top-level functions in a Bolt file.  Function
+definitions look just like _type_ and _path_ methods, except they can also accepts parameters.
+
+```javascript
+path /users/$userid is User {
+  read() = true;
+  write() = isUser($userid);
+}
+
+type User {
+  name: String,
+  age: Number | Null
+}
+
+// Define isUser() function to test if the given user id
+// matches the currently signed-in user.
+isUser(uid) = auth != null && auth.uid == uid;
+```
+
+```JSON
+{
+  "rules": {
+    "users": {
+      "$userid": {
+        ".validate": "newData.hasChildren(['name'])",
+        "name": {
+          ".validate": "newData.isString()"
+        },
+        "age": {
+          ".validate": "newData.isNumber() || newData.val() == null"
+        },
+        "$other": {
+          ".validate": "false"
+        },
+        ".read": "true",
+        ".write": "auth != null && auth.uid == $userid"
+      }
+    }
+  }
+}
+```
+
+# Bolt Cookbook
+
+The rest of this guide will provide sample recipes to solve typical problems that developers
+face in securing their Firebase databases.
+
+## Dealing with Timestamps
+
+You can write timestamps (Unix time in milliseconds) in Firebase data and ensure that whenever
+a time is written, it exactly matches the (trusted) server time (independent of the clock on
+the client device).
+
+```javascript
+path /posts/$id is Post;
+
+type Post {
+  // Make sure that the only value allowed to be written is now.
+  validate() = this.modified == now;
+
+  message: String,
+  modified: Number
+}
+```
+
+Each time the Post is written, modified must be set to the current time (using
+[ServerValue.TIMESTAMP](https://www.firebase.com/docs/web/api/servervalue/timestamp.html)).
+
+A handy way to express this is to use a user-defined type for the Modified timestamp:
+
+```javascript
+path /posts/$id is Post {
+  read() = true;
+  write() = true;
+}
+
+type Post {
+  message: String,
+  modified: Modified
+}
+
+type Modified extends Number {
```

Comment:
This is a very nice use-case for a custom type. I'm not sold on the name Modified, but have a hard time coming up with a better name.